src/db.rs

```diff
@@ -31,20 +32,20 @@ use std::ffi::{CStr, CString};
 use std::fmt;
 use std::fs;
 use std::iter;
-use std::marker::PhantomData;
 use std::path::Path;
 use std::path::PathBuf;
 use std::ptr;
 use std::slice;
 use std::str;
+use std::sync::Arc;
 use std::sync::RwLock;
 use std::time::Duration;
 
 // Marker trait to specify single or multi threaded column family alternations for DB
 // Also, this is minimum common API sharable between SingleThreaded and
 // MultiThreaded. Others differ in self mutability and return type.
 pub trait ThreadMode {
-    fn new(cf_map: BTreeMap<String, ColumnFamily>) -> Self;
+    fn new(cf_map: BTreeMap<String, *mut crate::ffi::rocksdb_column_family_handle_t>) -> Self;
```

Comment:
According to `git grep`, it looks like those C types tend to be prefixed with `ffi::`. So, I followed the convension instead of `use` just for `rocksdb_column_family_handle_t` (201d718):

```
$ git grep 'rocksdb_.*_t[^_a-z0-9]' src/
src/backup.rs:    inner: *mut ffi::rocksdb_backup_engine_t,
src/backup.rs:    inner: *mut ffi::rocksdb_options_t,
src/backup.rs:    inner: *mut ffi::rocksdb_restore_options_t,
src/backup.rs:        let be: *mut ffi::rocksdb_backup_engine_t;
src/checkpoint.rs:    inner: *mut ffi::rocksdb_checkpoint_t,
src/checkpoint.rs:        let checkpoint: *mut ffi::rocksdb_checkpoint_t;
src/column_family.rs:    pub(crate) inner: *mut ffi::rocksdb_column_family_handle_t,
src/column_family.rs:    pub(crate) inner: *mut ffi::rocksdb_column_family_handle_t,
src/column_family.rs:    pub(crate) inner: *mut ffi::rocksdb_column_family_handle_t,
src/column_family.rs:fn destroy_handle(handle: *mut ffi::rocksdb_column_family_handle_t) {
src/column_family.rs:    fn inner(&self) -> *mut ffi::rocksdb_column_family_handle_t;
src/column_family.rs:    fn inner(&self) -> *mut ffi::rocksdb_column_family_handle_t {
src/column_family.rs:    fn inner(&self) -> *mut ffi::rocksdb_column_family_handle_t {
src/column_family.rs:    fn inner(&self) -> *mut ffi::rocksdb_column_family_handle_t {
src/compaction_filter_factory.rs:    unsafe fn from_raw(ptr: *mut ffi::rocksdb_compactionfiltercontext_t) -> Self {
src/compaction_filter_factory.rs:    context: *mut ffi::rocksdb_compactionfiltercontext_t,
src/db.rs:    fn new(cf_map: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>) -> Self;
src/db.rs:    fn new(cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>) -> Self {
src/db.rs:    fn new(cfs: BTreeMap<String, *mut ffi::rocksdb_column_family_handle_t>) -> Self {
src/db.rs:        cfopts: &[*const ffi::rocksdb_options_t],
src/db.rs:        cfhandles: &mut Vec<*mut ffi::rocksdb_column_family_handle_t>,
src/db.rs:    ) -> Result<*mut ffi::rocksdb_column_family_handle_t, Error> {
src/db.rs:            // rocksdb_wal_readoptions_t does not appear to have any functions
src/db.rs:            let opts: *const ffi::rocksdb_wal_readoptions_t = ptr::null();
src/db.rs:        cf_inner: *mut ffi::rocksdb_column_family_handle_t,
src/db_iterator.rs:    inner: *mut ffi::rocksdb_iterator_t,
src/db_iterator.rs:        cf_handle: *mut ffi::rocksdb_column_family_handle_t,
src/db_iterator.rs:        cf_handle: *mut ffi::rocksdb_column_family_handle_t,
src/db_iterator.rs:    pub(crate) inner: *mut ffi::rocksdb_wal_iterator_t,
src/db_options.rs:fn new_cache(capacity: size_t) -> *mut ffi::rocksdb_cache_t {
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_cache_t,
src/db_options.rs:    inner: *mut ffi::rocksdb_env_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_options_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_writeoptions_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_flushoptions_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_block_based_table_options_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_readoptions_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_cuckoo_table_options_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_ingestexternalfileoptions_t,
src/db_options.rs:        unsafe { ffi::rocksdb_block_based_options_set_data_block_index_type(self.inner, index_t) }
src/db_options.rs:            .map(|path| path.inner as *const ffi::rocksdb_dbpath_t)
src/db_options.rs:            ffi::rocksdb_readoptions_set_readahead_size(self.inner, v as size_t);
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_fifo_compaction_options_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_universal_compaction_options_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_compactoptions_t,
src/db_options.rs:    pub(crate) inner: *mut ffi::rocksdb_dbpath_t,
src/db_pinnable_slice.rs:    ptr: *mut ffi::rocksdb_pinnableslice_t,
src/db_pinnable_slice.rs:    pub(crate) unsafe fn from_c(ptr: *mut ffi::rocksdb_pinnableslice_t) -> DBPinnableSlice<'a> {
src/perf.rs:    pub(crate) inner: *mut ffi::rocksdb_perfcontext_t,
src/perf.rs:    inner: *mut ffi::rocksdb_memory_usage_t,
src/perf.rs:    inner: *mut ffi::rocksdb_memory_consumers_t,
src/slice_transform.rs:    pub inner: *mut ffi::rocksdb_slicetransform_t,
src/snapshot.rs:    pub(crate) inner: *const ffi::rocksdb_snapshot_t,
src/sst_file_writer.rs:    pub(crate) inner: *mut ffi::rocksdb_sstfilewriter_t,
src/sst_file_writer.rs:    inner: *mut ffi::rocksdb_envoptions_t,
src/sst_file_writer.rs:    fn create_raw(opts: &Options, env_opts: &EnvOptions) -> *mut ffi::rocksdb_sstfilewriter_t {
src/write_batch.rs:    pub(crate) inner: *mut ffi::rocksdb_writebatch_t,

```